index.js

```diff
@@ -168,21 +180,26 @@ module.exports = class RandomAccessFile extends RandomAccessStorage {
       fs.rmdir(dir, onrmdir)
     }
   }
-}
 
-function readEmpty (req) {
-  req.callback(null, Buffer.alloc(0))
-}
+  _lock (mode, wait, cb) {
+    if (!fsctl) return cb(null)
+
+    const opts = { exclusive: mode === READWRITE }
+
+    if (wait) fsctl.lock(this.fd, opts).then(cb, cb)
```

Comment:
This is unsafe because if there's an error inside `cb` then it will be caught by the implicit `try/catch` from `Promise#then`. Node's `util.callbackify` is useful here

Reply:
@juliangruber actually wont! then(cb, cb) is very cool like that. then(cb).catch(cb) is unsafe tho

Reply:
In the browser it does 🤔 

```
const cb = () => { throw new Error('oops') }
Promise.resolve('hi').then(cb, cb)
Promise {<rejected>: Error: oops
    at cb (<anonymous>:1:26)}
VM808:1 Uncaught (in promise) Error: oops
    at cb (<anonymous>:1:26)
```

Reply:
same behavior with `p.then(cb).catch(cb)`, but not when you escape the stack:

```js
p.then(() => setTimeout(cb), () => setTimeout(cb))
Promise {<fulfilled>: 485}
VM808:1 Uncaught Error: oops
    at cb (<anonymous>:1:26)
```

Reply:
This in node logs "it was rejected" twice:

```js
const cb = () => { throw new Error('oops') }
Promise.resolve('hi').then(cb, cb).catch(() => console.log('it was rejected'))
Promise.resolve('hi').then(cb).catch(cb).catch(() => console.log('it was rejected'))
```

Unless I misunderstand?

Reply:
That’s a feature, it makes the promise unhandled as is expected

Reply:
strong disagree, callback error is a panic. crash the app

Reply:
otherwise the callback would have called cb(err)

Reply:
> strong disagree, callback error is a panic. crash the app

I agree, if we use `.then(() => setTimeout(cb))` then a error inside `cb` will lead to an unhandled exception, which is the process panic. If we don't use `setTimeout` or `util.callbackify` then we will have an unhandled rejection which doesn't always cause a panic.